File: ponio/examples/curtiss_hirschfelder.cpp
// Copyright 2022 PONIO TEAM. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

// NOLINTBEGIN(misc-include-cleaner)

#include <cmath>
#include <filesystem>
#include <string>

#include <ponio/observer.hpp>
#include <ponio/problem.hpp>
#include <ponio/runge_kutta.hpp>
#include <ponio/solver.hpp>

// NOLINTEND(misc-include-cleaner)

/*
solve Curtiss and Hirschfelder problem:

$$
    \begin{aligned}
        \dot{y} =  k(\cos(t) - y) \\
        y(0) = y_0
    \end{aligned}
$$

$y_0 = 2$
*/

int
main()
{
    std::string const dirname = "ch_data";

    using state_t = double;

    double const tf = 2.0;
    double const dt = 0.05;

    double const k = 50;

    auto ch_pb = ponio::implicit_problem(
        [=]( double t, double y, double& dy )
        {
            dy = k * ( std::cos( t ) - y );
        },
        [=]( double /* t */, double /* y */ )
        {
            return -k;
        } );

    state_t const y_0 = 2.0;

    { // simple use case with explicit method
        auto filename = std::filesystem::path( dirname ) / "sol_rk_33_ralston_cst.dat";
        auto obs      = ponio::observer::file_observer( filename );

        ponio::solve( ch_pb, ponio::runge_kutta::rk_33_ralston(), y_0, { 0., tf }, dt, obs );
    }

    { // simple use case with diagonal implicit method
        auto filename = std::filesystem::path( dirname ) / "sol_dirk23_cst.dat";
        auto obs      = ponio::observer::file_observer( filename );

        ponio::solve( ch_pb, ponio::runge_kutta::dirk23().newton_tol( 1e-5 ).newton_max_iter( 1000 ), y_0, { 0., tf }, dt, obs );
    }

    { // example of time loop with while loop controlled by user
        auto filename = std::filesystem::path( dirname ) / "sol_rk_33_ralston.dat";
        auto obs      = ponio::observer::file_observer( filename );

        auto sol_range = ponio::make_solver_range( ch_pb, ponio::runge_kutta::rk_33_ralston(), y_0, { 0., 0.464, tf }, dt );
        auto it_sol    = sol_range.begin();

        while ( it_sol->time < tf )
        {
            obs( it_sol->time, it_sol->state, it_sol->time_step );

            // pseudo adaptive time-step method
            if ( it_sol->time < 0.5 )
            {
                it_sol->time_step = 0.01;
                ++it_sol;
            }
            else
            {
                it_sol->time_step = 0.05;
                ++it_sol;
            }
        }
        obs( it_sol->time, it_sol->state, tf - it_sol->time ); // to save iteration where it_sol->time == tf
    }

    { // example of time loop with for loop on range
        auto filename = std::filesystem::path( dirname ) / "sol_rk54_6m.dat";
        auto obs      = ponio::observer::file_observer( filename );

        auto sol_range = ponio::make_solver_range( ch_pb, ponio::runge_kutta::rk54_6m(), y_0, { 0., tf }, dt );

        for ( auto ui : sol_range )
        {
            obs( ui.time, ui.state, ui.time_step );
        }
    }

    return 0;
}

File: ponio/examples/curtiss_hirschfelder.py
# Copyright 2022 PONIO TEAM. All rights reserved.
# Use of this source code is governed by a BSD-style
# license that can be found in the LICENSE file.

#!/usr/bin/env python

import subprocess
import os
import argparse

import matplotlib.pyplot as plt
import matplotlib.cm as cm
import numpy as np

name = "curtiss_hirschfelder"
data_dir = "ch_data"
img_dir = data_dir

parser = argparse.ArgumentParser(
    description=f"Compile, launch and plot results of example `{name}`")
parser.add_argument('--only-save', action='store_true',
                    help="Just save output in img directory")

arguments = parser.parse_args()

if arguments.only_save:
    img_dir = os.path.join("img", name)
    os.makedirs(img_dir, exist_ok=True)


make = subprocess.Popen(["make", name])
make.wait()

args = [os.path.join(".", name)]
process = subprocess.Popen(args)
process.wait()

methods = {
    "rk_33_ralston_cst": "RK(3,3) Ralston fixed $\\Delta t$",
    "dirk23_cst": "DIRK(2, 3) fixed $\\Delta t$",
    "rk_33_ralston": "RK(3,3) Ralston",
    "rk54_6m": "RK5(4) 6m"
}

fig, axs = plt.subplots(2)
# axs[0].axvline(0.464, linewidth=1, linestyle="--", color="grey")
# axs[1].axvline(0.464, linewidth=1, linestyle="--", color="grey")

line_style = "+-"
for tag, meth in methods.items():
    data = np.loadtxt(os.path.join(data_dir, f"sol_{tag}.dat"))
    t, y, dt = data[:, 0], data[:, 1], data[:, 2]

    axs[0].plot(t, y, line_style, label=f"{meth}")
    axs[1].plot(t, dt, line_style, label=f"{meth}")
    line_style = "x-"

axs[0].set_ylabel("$y$")
axs[1].set_ylabel("$\\Delta t$")
axs[1].set_xlabel('time')

plt.legend(loc="lower right")

plt.savefig(os.path.join(img_dir, "01.png"), dpi=200)

if not arguments.only_save:
    fig.suptitle("Solution of Curtiss Hirschfelder problem")
    plt.show()
